src/api.js

```diff
@@ -17,33 +18,376 @@ const getMessages = args => {
+      const index = findTargetIndex(id, composition_msgs);
+      if (composition_msgs[index].groupId === groupId) {
+        group.messages.push(composition_msgs.splice(index, 1)[0]);
+      } else {
+        if (composition_msgs[index].groupId) {
+          if (!inner_groups[composition_msgs[index].groupId]) {
+            inner_groups[composition_msgs[index].groupId] = [];
+          }
+          inner_groups[composition_msgs[index].groupId].push(composition_msgs.splice(index, 1)[0]);
+        } else {
+          group.messages.push(assemblePipeline(index, composition_msgs));
+        }
+      }
+    });
+    Object.values(inner_groups).forEach(inner_grp => {
+      const grp = assembleGroup(
+        inner_grp.map(msg => msg.messageId),
+        inner_grp
+      );
+      composition_msgs.push(grp);
+      group.messages.push(assemblePipeline(composition_msgs.length - 1, composition_msgs));
+    });
+    if (group.messages[0].pipeTarget) {
+      group.pipeTarget = group.messages[0].pipeTarget;
+    }
+    group.messageId = group.messages[0].messageId;
+    return addDetails(group);
+  }
+
+  function assembleComposition(composition_index, composition_msgs) {
+    // finding the index of one of the first messages of a composition
+    let index = 0;
+    let prev_index = findPreviousElement(index, composition_msgs);
+    while (prev_index !== -1) {
+      index = prev_index;
+      prev_index = findPreviousElement(index, composition_msgs);
+    }
+    if (composition_msgs[index].groupId) {
+      const group_indexes = [];
+      composition_msgs.forEach((message, index_message) => {
+        if (message.groupId === composition_msgs[index].groupId) {
+          group_indexes.push(index_message);
+        }
+      });
+      const group_ids = group_indexes.map(index_message => {
+        let prev_index = findPreviousElement(index_message, composition_msgs);
+        while (prev_index !== -1 && composition_msgs[prev_index].pipeTarget.length === 1) {
+          index_message = prev_index;
+          prev_index = findPreviousElement(index_message, composition_msgs);
+        }
+        return composition_msgs[index_message].messageId;
+      });
+      const group = assembleGroup(group_ids, composition_msgs);
+      if (group.pipeTarget) {
+        composition_msgs.push(group);
+        return assembleComposition(composition_msgs.length - 1, composition_msgs);
+      } else {
+        return group;
+      }
+    } else {
+      const pipeline = assemblePipeline(index, composition_msgs);
+      if (pipeline.groupId) {
+        composition_msgs.push(pipeline);
+        return assembleComposition(composition_msgs.length - 1, composition_msgs);
+      }
+      return pipeline;
+    }
+  }
+
+  Object.entries(compositions).forEach(composition_messages => {
+    const composition = assembleComposition(0, composition_messages[1]);
+    messages.push(composition);
+  });
+
+  return { messages: messages, loadDateTime: loadDateTime };
+}
```

Comment:
ok this is quite long, I will read all this later. Put it in his own file and add a docstring to each module so we know what it is used for.